Feature/implement-converter

tytianiuk opened the pull request:
I implemented :
- converting CSV and JS to JSON
- uploading files
- copying and downloading converted files

tytianiuk committed:
Implement converting CSV and JS files

tytianiuk committed:
Add uploading files, downloading and copying results

ihornovykov commented on src/app/_components/convertor.tsx:
```diff
@@ -1,61 +1,179 @@
+        setText('')
+    }
+
+    const handleConvert = (format: string | null) => {
+        if (format === 'csv') {
+            const parsedData = parserCSV(text)
+            if (parsedData) {
+                const { keys, values } = parsedData
+                setJson(convertToJSON(keys, values))
+            } else setJson('Error: invalid CSV data')
+        } else if (format === 'js') {
+            const parsedData = parserJS(text)
+            if (parsedData) {
+                setJson(parsedData)
+            } else setJson('Error: invalid JavaScript data')
+        } else return setJson('Error: select file type')
+    }
```
я думаю тут краще винести цю функцію в папку utils або зробити кастомний хук(так навіть краще) і замість великої к-ті if else зробити об'єкт по типу:

```js
const converters = {
  js: {
    converter: parserJS // Не викликаючи
    error: 'Error: invalid JavaScript data'  
  } // і тд
}
```

а потім вже звертатись до властивості converter і викликати з параметром

ну і ще можна перейменувати цю функцію як конвертер

ihornovykov commented on src/app/_components/convertor.tsx:
```diff
@@ -1,61 +1,179 @@
+        } else return setJson('Error: select file type')
+    }
+
+    const parserCSV = (data: string) => {
+        const lines = data.split('\n')
+        const keys = lines[0].split(',').map((key) => key.trim())
+        const values = lines
+            .slice(1)
+            .map((row) => row.split(',').map((value) => value.trim()))
+
+        return checkValues(keys, values) ? null : { keys, values }
+    }
+
+    const parserJS = (data: string) => {
+        try {
+            return eval(`${data}`)
+        } catch (error) {
+            console.error('Error:', error)
+            return null
+        }
+    }
```
оці функції в папку utils треба перенести

ihornovykov commented on src/app/_components/convertor.tsx:
```diff
@@ -1,61 +1,179 @@
+        }
+    }
+
+    const convertToJSON = (keys: string[], values: string[][]) => {
+        const json: any[] = []
+
+        values.forEach((row) => {
+            const newObject: { [key: string]: string } = {}
+            row.map((value, index) => {
+                newObject[keys[index]] = value
+            })
+            json.push(newObject)
+        })
+
+        return json
+    }
```
і це теж

tytianiuk committed:
Fix issues from code review
- Add custom hook: useConverter
- Put functions to utils folder
- Remove dublicate code

tytianiuk committed:
Implement "drag and drop"

tytianiuk committed:
Add display of the uploaded file
- add option to delete the uploaded file
- change download button icon

tytianiuk committed:
Add option to convert XML to JSON
